dev-cmd/bottle: Support adding root_url specs to bottle spec

cwegrzyn opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This allows additional root_url specs to be provided with a custom root_url when `brew bottle` updates a formula. This addresses the feature request raised in #5424. Specifically, it makes it easier to support private taps that may need additional authentication to fetch artifacts.

As an alternative to this, I considered implementing a flag to accept a complete replacement bottle ERB template. This had the advantage of simplifying maintenance of this functionality from a homebrew maintainer perspective, while allowing for further arbitrary customizations of the bottle block. On the other hand, if the bottle template is likely to change significantly, users of this functionality could wind up with automated processes generating broken formulae until they update their bottle template, which also didn't seem awesome. But I'd be happy to implement this strategy if preferred-- either would work for me at the moment.

MikeMcQuaid commented on Library/Homebrew/test/dev-cmd/bottle_spec.rb:
```diff
@@ -415,6 +415,90 @@ def install
+    end
+  end
+
+  describe "--merge with --root-url-specs", :integration_test do
```
Would rather not add another integration test for this. They are really slow and this is very niche behaviour.

cwegrzyn replied:
Makes sense! Doesn’t look like there are any direct tests of ‘bottle_output’ at present. Would you suggest adding some or not bothering with tests here?

MikeMcQuaid commented:
> This allows additional root_url specs to be provided with a custom root_url when `brew bottle` updates a formula. This addresses the feature request raised in #5424. Specifically, it makes it easier to support private taps that may need additional authentication to fetch artifacts.

Thanks for the PR @cwegrzyn! I cannot promise yet it'll be merged. A few questions:
- where/how are you uploading the bottles you created?
- what are some examples you are passing to use this argument?

cwegrzyn commented:
> * where/how are you uploading the bottles you created?

Currently uploading to GitHub releases in a private repo. With a few small tweaks (and this PR), I’ve put together a GitHub actions workflow based roughly on the output of tap-new that works for a private repo.

> * what are some examples you are passing to use this argument?

currently just `--root-url-specs="using: GitHubPrivateRepositoryReleaseDownloadStrategy"` (where that strategy is pulled in via require_relative in the formulae). I considered making the argument just `—download-strategy` like in the original issue, but the underlying code seems to support arbitrary url specs, so I figured might as well just make it generic in case other args might prove useful for someone.

MikeMcQuaid commented:
> Currently uploading to GitHub releases in a private repo.

Are you using `brew pr-upload` or similar for that? It feels like that may be another place to handle this.

> I considered making the argument just `—download-strategy` like in the original issue, but the underlying code seems to support arbitrary url specs, so I figured might as well just make it generic in case other args might prove useful for someone.

I politely disagree. I'd suggest this is more useful/obvious if it's more specific i.e. only passing `using:` here.

cwegrzyn commented:
> Are you using `brew pr-upload` or similar for that? It feels like that may be another place to handle this.

Hmm, I am using `brew pr-upload` by way of `pr-pull` in my workflow. Are you thinking there is an alternative way to handle this purely in pr-upload? I added the pass through of this arg from `pr-pull` to `pr-upload` to `bottle` because it seemed like I needed to adjust the bottle block before the commit happened. But happy to switch it up!

> I politely disagree. I'd suggest this is more useful/obvious if it's more specific i.e. only passing `using:` here.

Fair enough! I can switch it out to a download strategy specific argument. That certainly suffices for my current use case. Will update the PR a little later today.

MikeMcQuaid commented:
> Hmm, I am using `brew pr-upload` by way of `pr-pull` in my workflow. Are you thinking there is an alternative way to handle this purely in pr-upload? I added the pass through of this arg from `pr-pull` to `pr-upload` to `bottle` because it seemed like I needed to adjust the bottle block before the commit happened. But happy to switch it up!

Note purely in `pr-upload` but given they both taken `--root-url=` I wonder if they need adjusted too?



> Fair enough! I can switch it out to a download strategy specific argument. That certainly suffices for my current use case. Will update the PR a little later today.

Yeh, this feels like the right call. Thanks! ❤️

cwegrzyn commented:
Okay! Rewrote in terms of a `--download-strategy` option. I also replaced the integration test with a simpler unit test. How do things look now?

As for `brew pr-upload`, I took a look at uses of root_url there. Besides passing it along to `brew bottle`, the other usage I could identify is to determine where to upload the packages (e.g., bintray vs github packages vs github releases). Since the github releases uploader is already compatible with a private repo, I didn't need to change any code here for my purposes (where my custom download strategy is for downloading from private repo github releases). That said, if someone were using a custom download strategy to store bottles somewhere altogether different, they'd still currently be on their own for the upload portion. Do you think it would be helpful to add a brief note to the `--download-strategy` help text in `brew pr-upload` that it doesn't change how the packages are _uploaded_?

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -16,7 +16,9 @@
     <% if [HOMEBREW_BOTTLE_DEFAULT_DOMAIN.to_s,
            "#{HOMEBREW_BOTTLE_DEFAULT_DOMAIN}/bottles"].exclude?(root_url) %>
-    root_url "<%= root_url %>"
+    root_url "<%= root_url %>"<% unless download_strategy.blank? %>,
```
```suggestion
    root_url "<%= root_url %>"<% if download_strategy.present? %>,
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -75,6 +77,8 @@ def bottle_args
              description: "Specify a committer name and email in `git`'s standard author format."
       flag   "--root-url=",
              description: "Use the specified <URL> as the root of the bottle's URL instead of Homebrew's default."
+      flag   "--download-strategy=",
```
```suggestion
      flag   "--root-url-download-strategy=",
```
or
```suggestion
      flag   "--root-url-using=",
```

cwegrzyn replied:
Ah, yeah, makes a lot of sense. Went with `--root-url-using` with the logic that it should be directly obvious to someone who is trying to get one of these commands to replicate a `using` clause they've probably already tried out by hand.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/pr-pull.rb:
```diff
@@ -61,6 +61,8 @@ def pr_pull_args
              description: "Target tap repository (default: `homebrew/core`)."
       flag   "--root-url=",
              description: "Use the specified <URL> as the root of the bottle's URL instead of Homebrew's default."
+      flag   "--download-strategy=",
```
Same here and below

MikeMcQuaid reviewed (commented):
Looking good! One suggested naming tweak.

cwegrzyn commented:
Okay, all set! Thanks for bearing with me on this!

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -75,6 +77,8 @@ def bottle_args
       flag   "--root-url=",
              description: "Use the specified <URL> as the root of the bottle's URL instead of Homebrew's default."
+      flag   "--root-url-using=",
+             description: "Use the specified download strategy for the root_url in the generated DSL via `using:' arg"
```
```suggestion
             description: "Use the specified download strategy class for downloading the bottle's URL instead of Homebrew's default."
```

(and same below)

cwegrzyn commented:
Definitely a better description, thanks!

MikeMcQuaid commented:
Nice work @cwegrzyn! Can you rebase this and run `brew style --fix`? For `description: "Use the specified download strategy class for downloading the bottle's URL instead of Homebrew's default."`: split the string and use `\` to put it on another line.

cwegrzyn commented:
Doh! Thought I had run a brew style before pushing up last time, oh well. All set now!

MikeMcQuaid commented:
Thanks so much for your first contribution! Without people like you submitting PRs we couldn't run this project. You rock, @cwegrzyn!
